### DISCUSSION TITLE
[Dev Tools] `cast run` not working for era sepolia transaction

### DISCUSSION THREAD

**Discussion by coffee-the-dev** [zkSync-Community-Hub/zksync-developers#806] [Dev Tooling] [answered]
### Environment

Testnet

### Select the Dev Tool you are using

Foundry

### Provide the version of the tool (if applicable)

cast 0.0.2 (15b4758 2024-12-03T00:28:46.765147000Z)

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

I am trying to trace a zksync era sepolia transaction to get a better sense of the transaction flow that occured using the following command:
```
cast run --rpc-url https://sepolia.era.zksync.dev 0x33bf84fc5a25138218d365b9b0a551d10e64e70d8923a09479089cd0cd48930d
```

I get the following error when running this command:
```
Executing previous transactions from the block.
Error: backend: failed while inspecting; transaction validation error: lack of funds (0) for max fee (24597360000000); lack of funds (0) for max fee (24597360000000);
```

I would expect this to work as it is supposed to simply replay transactions from the block and then output the traces of the transaction specified.

A functional example of this behavior can be seen by using `cast run` against a mainnet eth transaction:
```
cast run --rpc-url https://rpc.ankr.com/eth 0x328eb7e21e0bf3d21e6bb60581a7efced4e17e82c1b93e0d6b64d798f957e938
```
Result:
```
Executing previous transactions from the block.
Traces:
  [86689] 0x99B1817acb40E76C309e26B2Face9da9EFf55317::5a7b05e8(0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000008800000000000000000000000000000000000000000000000010f17f7cc6c7d68000000000000000000000000000000000000000000000327ff664d29242a3d033800000000000000000000000e331de28cd81b768c19a366b0e4e4675c45ec2dadfc33cfe4438e623d46d4a06272dc677838bf94e00000000000000000093c16a28a863ba6a4086c5000000000000000000000000000000000000000000000000)
    ├─ [2696] 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA::slot0() [staticcall]
    │   └─ ← [Return] 0x00000000000000000000000000000000000000000093c16a28a863ba6a4086c5fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe23d10000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004c000000000000000000000000000000000000000000000000000000000000004c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001
    ├─ [76383] 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA::swap(0xDfC33cfe4438e623D46d4a06272Dc677838bf94E, false, 1220897147881772672 [1.22e18], 1461446703485210103287273052203988822378723970341 [1.461e48], 0x000000000000000000000000ab95799727dfe3431bf4b52d2bafc6f8a6fac2c3)
    │   ├─ [13999] 0x55296f69f40Ea6d20E478533C15A6B08B654E758::transfer(0xDfC33cfe4438e623D46d4a06272Dc677838bf94E, 238481200000000000167989 [2.384e23])
    │   │   ├─ emit Transfer(src: 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA, dst: 0xDfC33cfe4438e623D46d4a06272Dc677838bf94E, wad: 238481200000000000167989 [2.384e23])
    │   │   └─ ← [Stop]
    │   ├─ [2534] WETH9::balanceOf(0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA) [staticcall]
    │   │   └─ ← [Return] 275690035847047211982 [2.756e20]
    │   ├─ [15651] 0x99B1817acb40E76C309e26B2Face9da9EFf55317::uniswapV3SwapCallback(-238481200000000000167989 [-2.384e23], 1220897147881772672 [1.22e18], 0x000000000000000000000000ab95799727dfe3431bf4b52d2bafc6f8a6fac2c3)
    │   │   ├─ [266] 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA::token0() [staticcall]
    │   │   │   └─ ← [Return] 0x00000000000000000000000055296f69f40ea6d20e478533c15a6b08b654e758
    │   │   ├─ [308] 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA::token1() [staticcall]
    │   │   │   └─ ← [Return] 0x000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
    │   │   ├─ [251] 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA::fee() [staticcall]
    │   │   │   └─ ← [Return] 0x0000000000000000000000000000000000000000000000000000000000000bb8
    │   │   ├─ [13025] WETH9::transferFrom(0xAb95799727DFe3431BF4B52D2Bafc6F8a6fAC2C3, 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA, 1220897147881772672 [1.22e18])
    │   │   │   ├─ emit Transfer(src: 0xAb95799727DFe3431BF4B52D2Bafc6F8a6fAC2C3, dst: 0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA, wad: 1220897147881772672 [1.22e18])
    │   │   │   └─ ← [Return] true
    │   │   └─ ← [Stop]
    │   ├─ [534] WETH9::balanceOf(0xE331DE28cd81B768C19A366b0e4e4675c45eC2dA) [staticcall]
    │   │   └─ ← [Return] 276910932994928984654 [2.769e20]
    │   ├─ emit Swap(param0: 0x99B1817acb40E76C309e26B2Face9da9EFf55317, param1: 0xDfC33cfe4438e623D46d4a06272Dc677838bf94E, param2: -238481200000000000167989 [-2.384e23], param3: 1220897147881772672 [1.22e18], param4: 179364243816031445391199306 [1.793e26], param5: 130539702943625079783855 [1.305e23], param6: -121820 [-1.218e5])
    │   └─ ← [Return] 0xffffffffffffffffffffffffffffffffffffffffffffcd7fe8829f8cc9456fcb00000000000000000000000000000000000000000000000010f17f7cc6c7d680
    └─ ← [Return] 0x000000000000000000000000000000000000000000003280177d607336ba9035


Transaction successfully executed.
Gas used: 109513
```

**Comment by alhenaai** [comment]
The error you're encountering with `cast run` on the ZKsync Era Sepolia testnet is due to a lack of funds for the transaction fee. This issue arises because the account you're using to replay the transaction does not have sufficient balance to cover the maximum fee specified in the transaction.

Here are a few steps you can take to resolve this issue:

1. **Ensure Sufficient Balance**: Make sure that the account you are using has enough funds to cover the transaction fees. You might need to deposit additional funds into this account.

2. **Check the RPC URL**: Verify that the RPC URL (`https://sepolia.era.zksync.dev`) is correct and that the node is fully synced and operational.

3. **Transaction Details**: Double-check the transaction hash and ensure it is correct and belongs to the Sepolia testnet.

4. **Tool Version**: Ensure that your version of `cast` is compatible with ZKsync Era transactions. Sometimes, updates or specific configurations might be required to handle newer network features or changes.

If you continue to face issues, consider reaching out for more specific support or guidance through the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) where you can get more detailed assistance tailored to your situation.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by dutterbutter** [comment, marked as answer]
Hey @coffeexcoin 👋 

`cast run` is currently [not supported](https://foundry-book.zksync.io/supported-commands/). You can still `replay_tx` via `era-test-node` (soon to be `anvil-zksync`). You can do so like so:

```
./target/release/era_test_node --show-calls=all replay_tx 0x33bf84fc5a25138218d365b9b0a551d10e64e70d8923a09479089cd0cd48930d --network sepolia-testnet
```

Here is a preview of new output that will be shipped in the `anvil-zksync` update from the `db/fix-and-improve-debugging-logs` branch 😁 for tx hash: `0x33bf84fc5a25138218d365b9b0a551d10e64e70d8923a09479089cd0cd48930d` on sepolia:

**Note:** I used `--show-calls=all` to display all system and user calls, it can be further filtered to just `user`. 

![replay_tx](https://github.com/user-attachments/assets/25bea074-6e92-41d5-b928-69d59d91b684)
